### PR TITLE
rls: Support multiple returned targets from RLS Server

### LINKED ISSUE #9236: rls: Support multiple returned targets
The RLS server can return multiple targets. We need to check all of them to find one that is connected. We shouldn't just assume the first will work:
https://github.com/grpc/grpc-java/blob/a206cda1a8e4a84aeef073e298d76a2692a6a002/rls/src/main/java/io/grpc/rls/CachingRlsLbClient.java#L518-L521

The multi-entry response was synced into grpc-java in https://github.com/grpc/grpc-java/pull/7046 . The code was updated to compile, but support the new feature.

### PR DESCRIPTION
**Pull request by larry-safran**
Pick the first target that is in a good state (READY or IDLE).  If none, then pick the first target in CONNECTING state.  If none, use the first target.

Fixes #9236

### CODE REVIEW

**Review by ejona86 — COMMENTED** [review]
I don't think this is the right approach. This will pin the result and not change until the cache entry is refreshed. It will also avoid attempting (create+wait for connect) earlier targets entirely if a later target is already READY.

File: rls/src/main/java/io/grpc/rls/CachingRlsLbClient.java [hunk 1]
```diff
@@ -57,9 +57,11 @@
 import io.grpc.rls.Throttler.ThrottledException;
 import io.grpc.stub.StreamObserver;
 import io.grpc.util.ForwardingLoadBalancerHelper;
+
```
Comment (ejona86) [inline rls/src/main/java/io/grpc/rls/CachingRlsLbClient.java:60]: Delete blank line.

File: rls/src/main/java/io/grpc/rls/CachingRlsLbClient.java [hunk 2]
```diff
@@ -591,7 +621,8 @@ boolean isStaled(long now) {
     @Override
     void cleanup() {
       synchronized (lock) {
-        refCountedChildPolicyWrapperFactory.release(childPolicyWrapper);
+        for (ChildPolicyWrapper policyWrapper : childPolicyWrappers)
```
Comment (ejona86) [inline rls/src/main/java/io/grpc/rls/CachingRlsLbClient.java:624]: [Our style guide](https://google.github.io/styleguide/javaguide.html#s4.1.1-braces-always-used) always uses braces for `for` blocks.

File: rls/src/main/java/io/grpc/rls/CachingRlsLbClient.java [hunk 3]
```diff
@@ -563,9 +573,29 @@ void maybeRefresh() {
...
+        }
+      }
+
+      throw new RuntimeException("Target not found:" + target);
+    }
+
     @Nullable
     ChildPolicyWrapper getChildPolicyWrapper() {
-      return childPolicyWrapper;
+      if (childPolicyWrappers == null || childPolicyWrappers.isEmpty()) {
```
Comment (ejona86) [inline rls/src/main/java/io/grpc/rls/CachingRlsLbClient.java:589]: How could the wrappers be null?

We should validate that there are targets when constructing the DataCacheEntry. It looks like that is safe to do in the constructor as all calls to `new DataCacheEntry()` are in a try-catch. That would avoid passing `null` to the data plane.

Comment (larry-safran) [reply rls/src/main/java/io/grpc/rls/CachingRlsLbClient.java:589]: You are correct that wrappers can't be null and resp.targets() can't be null, but if the response target list is empty, then the wrappers would be also.  Is it legal for RLS to return an empty list of targets?

Comment (ejona86) [reply rls/src/main/java/io/grpc/rls/CachingRlsLbClient.java]: I don't think it matters. If we don't have a list of targets, whether it is allowed or not, either way we're going to fail the RPC.

File: rls/src/main/java/io/grpc/rls/CachingRlsLbClient.java [hunk 4]
```diff
@@ -509,16 +519,16 @@ final class DataCacheEntry extends CacheEntry {
...
-    private final ChildPolicyWrapper childPolicyWrapper;
+    private final List<ChildPolicyWrapper> childPolicyWrappers;
 
     // GuardedBy CachingRlsLbClient.lock
     DataCacheEntry(RouteLookupRequest request, final RouteLookupResponse response) {
       super(request);
       this.response = checkNotNull(response, "response");
-      // TODO(creamsoup) fallback to other targets if first one is not available
-      childPolicyWrapper =
+      checkState(!response.targets().isEmpty(), "No targets returned by RLS");
```
Comment (ejona86) [inline rls/src/main/java/io/grpc/rls/CachingRlsLbClient.java:528]: I'd feel much more comfortable if this were a checked exception. Unhandled exception in LB policies put the channel in panic mode which fails all RPCs on the channel for the remainder of its life.

**Review by ejona86 — APPROVED** [review]

**Comment by ejona86** [comment]
This PR is for "rls" not "core". The commit/PR title is incorrect.

**Pull request merged**